tools/approval-service/internal/service/ttleventcache.go

```diff
@@ -0,0 +1,152 @@
+package service
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+)
+
+// TTLEventCache is a thread-safe dedupe cache.
+// - TryAdd: "debounce-on-arrival" (first arrival wins, subsequent arrivals within the TTL window are rejected
+type TTLEventCache struct {
+	mu              sync.Mutex
+	items           map[string]time.Time // expiry time for an item; zero time reserved for in-progress marker
+	ttl             time.Duration
+	cleanupInterval time.Duration
+	stop            chan struct{}
+	done            chan struct{}
+	closed          bool
+}
+
+// Option configures TTLEventCache construction
+type Option func(*TTLEventCache) error
+
+// WithCleanupInterval allows overriding the tick interval (used by janitor for cleanup) via option.
+// Validates duration > 0.
+func WithCleanupInterval(interval time.Duration) Option {
+	return func(c *TTLEventCache) error {
+		if interval <= 0 {
+			return fmt.Errorf("invalid interval")
+		}
+		c.cleanupInterval = interval
+		return nil
+	}
+}
+
+// MakeTTLEventCache creates a new TTLEventCache.
+// ttl: how long to prevent re-processing after an arrival or after finish().
+// Use options for additional configuration (e.g., WithCleanupInterval).
+func MakeTTLEventCache(ttl time.Duration, opts ...Option) (*TTLEventCache, func() error, error) {
+	// set a logical default
+	if ttl <= 0 {
+		ttl = 15 * time.Second
+	}
+
+	// default tick: ttl/4 for moderate TTLs, minimum 1s
+	defaultCleanupInterval := func(ttl time.Duration) time.Duration {
+		if ttl > 4*time.Second {
+			return ttl / 4
+		}
+		return 1 * time.Second
+	}
+
+	c := &TTLEventCache{
+		items:           make(map[string]time.Time),
+		ttl:             ttl,
+		cleanupInterval: defaultCleanupInterval(ttl),
```

Comment:
this can be simplified to `cleanupInterval: max(ttl / 4, 1*time.Second)`, removing the inline function above.